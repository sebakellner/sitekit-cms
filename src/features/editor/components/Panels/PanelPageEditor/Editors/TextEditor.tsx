import { Box, Field, Input } from '@chakra-ui/react'

import type { PanelEditorBase } from '@features/editor/types/panelEditors.types'

const TextEditor = ({ title, value, onChange }: PanelEditorBase) => {
  return (
    <Box mb={4}>
      <Field.Root orientation="horizontal">
        <Field.Label fontSize="xs">{title}</Field.Label>
        <Input
          value={value}
          onChange={(e) => onChange(e.target.value)}
          variant="subtle"
        />
      </Field.Root>
    </Box>
  )
}

export default TextEditor
