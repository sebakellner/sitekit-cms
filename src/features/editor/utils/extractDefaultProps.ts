import { ComponentMetaSchema } from '@components/site/schemas/componentMeta.schema'
import type { ComponentMeta } from '@components/site/types'

function isDefaultProp(val: unknown): val is { default: unknown } {
  return typeof val === 'object' && val !== null && 'default' in val
}

export function extractDefaultProps(
  props: ComponentMeta['props'] | undefined,
  fallback: unknown = undefined
): Record<string, unknown> {
  if (!props) return {}
  const result = ComponentMetaSchema.shape.props.safeParse(props)
  if (!result.success) {
    console.error('Invalid props schema:', result.error)
    return {}
  }
  const validProps = result.data
  return Object.fromEntries(
    Object.entries(validProps).map(([key, val]) => [
      key,
      isDefaultProp(val) ? val.default : fallback,
    ])
  )
}
