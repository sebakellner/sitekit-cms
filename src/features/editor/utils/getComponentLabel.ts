import React from 'react'

export function getComponentLabel(
  children: React.ReactNode,
  fallback = 'Component'
): string {
  if (React.isValidElement(children)) {
    const type = children.type
    if (typeof type === 'function' && type.name) return type.name
  }
  return fallback
}
