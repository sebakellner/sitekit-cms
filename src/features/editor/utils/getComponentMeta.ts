import { ComponentMetaSchema } from '@components/site/schemas/componentMeta.schema'

type ModuleWithDefault = { default: unknown }

export function getComponentMeta(module: unknown) {
  if (
    module &&
    typeof module === 'object' &&
    'default' in module &&
    (module as ModuleWithDefault).default &&
    typeof (module as ModuleWithDefault).default === 'object'
  ) {
    const def = (module as ModuleWithDefault).default
    const result = ComponentMetaSchema.safeParse(def)
    if (result.success) {
      return result.data
    }
  }
  return { name: '', props: {} }
}
