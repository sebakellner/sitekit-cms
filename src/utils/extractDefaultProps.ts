import type { ComponentMeta } from '@components/site/types'

export function extractDefaultProps(
  props: ComponentMeta['props'] | undefined
): Record<string, unknown> {
  if (!props) return {}
  return Object.fromEntries(
    Object.entries(props).map(([key, val]) => [
      key,
      (val as { default: unknown }).default,
    ])
  )
}
