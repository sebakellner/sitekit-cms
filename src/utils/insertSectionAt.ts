import type { Section } from '@src/types'

export function insertSectionAtUtil(
  sections: Section[],
  setSections: (sections: Section[]) => void,
  id: string,
  index: number
) {
  const sectionId = String(id).startsWith('selector-')
    ? String(id).replace('selector-', '')
    : id
  const uniqueId = `${sectionId}-${Date.now()}`
  const newSection = {
    id: uniqueId,
    name: sectionId,
    props: {},
  }
  const newSections = [...sections]
  newSections.splice(index, 0, newSection)
  setSections(newSections)
}
