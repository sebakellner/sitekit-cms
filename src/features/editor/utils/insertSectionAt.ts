import { componentsRegistry } from '@registry/componentRegistry'
import { v4 as uuidv4 } from 'uuid'

import type { ComponentMeta } from '@components/site/types'
import { extractDefaultProps } from '@src/features/editor/utils/extractDefaultProps'
import type { Section } from '../schemas/section.schema'

export async function insertSectionAtUtil(
  sections: Section[],
  setSections: (sections: Section[]) => void,
  name: string,
  index: number
) {
  const loader = componentsRegistry[name as keyof typeof componentsRegistry]
  if (!loader) {
    console.warn(`Component "${name}" not found in registry.`)
    return
  }

  const metaModule = await loader()

  const meta = metaModule.default as ComponentMeta

  const defaultProps = extractDefaultProps(meta?.props)

  const newSection: Section = {
    id: uuidv4(),
    name,
    props: defaultProps,
  }

  const updatedSections = [
    ...sections.slice(0, index),
    newSection,
    ...sections.slice(index),
  ]

  setSections(updatedSections)
}
