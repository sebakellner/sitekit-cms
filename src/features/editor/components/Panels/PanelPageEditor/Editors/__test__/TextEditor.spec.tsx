import { describe, test, expect } from 'vitest'
import { render } from 'test/utils/render'
import type { PanelEditorBase } from '@features/editor/types/panelEditors.types'
import { screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import TextEditor from '../TextEditor'
import { useState } from 'react'

const setup = (overrides: Partial<PanelEditorBase<string>> = {}) => {
  const defaultProps: PanelEditorBase<string> = {
    title: 'Title',
    value: 'This is a text editor',
    onChange: () => {},
  }
  const props = { ...defaultProps, ...overrides }

  const ControlledTextEditor = () => {
    const [value, setValue] = useState(props.value)
    return <TextEditor {...props} value={value} onChange={setValue} />
  }

  return <ControlledTextEditor />
}

describe('TextEditor', () => {
  test('should match with snapshot', () => {
    const { container } = render(setup())
    expect(container).toMatchSnapshot()
  })

  test('should call onChange with updated value', async () => {
    render(setup())

    const input = screen.getByLabelText('Title')
    await userEvent.clear(input)
    expect(input).toHaveValue('') // Ahora sí debe estar vacío
    await userEvent.type(input, 'New Title')
    await userEvent.tab()

    await waitFor(() => {
      expect(input).toHaveValue('New Title')
    })
  })
})
