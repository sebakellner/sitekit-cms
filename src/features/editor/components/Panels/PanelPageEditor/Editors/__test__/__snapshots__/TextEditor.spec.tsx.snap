// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TextEditor > should match with snapshot 1`] = `
<div>
  <script
    nonce=""
  >
    ((e,i,s,u,m,a,l,h)=&gt;{let d=document.documentElement,w=["light","dark"];function p(n){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&a?m.map(f=&gt;a[f]||f):m;k?(d.classList.remove(...S),d.classList.add(a&&a[n]?a[n]:n)):d.setAttribute(y,n)}),R(n)}function R(n){h&&w.includes(n)&&(d.style.colorScheme=n)}function c(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(u)p(u);else try{let n=localStorage.getItem(i)||s,y=l&&n==="system"?c():n;p(y)}catch(n){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
  <div
    class="css-n33ttv"
  >
    <div
      class="chakra-field__root css-1anu6qw"
      data-part="root"
      data-scope="field"
      id="field::«r0»"
      orientation="horizontal"
      role="group"
    >
      <label
        class="chakra-field__label css-84rnco"
        data-part="label"
        data-scope="field"
        for="«r0»"
        id="field::«r0»::label"
      >
        Title
      </label>
      <input
        class="chakra-input css-1hens3x"
        data-part="input"
        data-scope="field"
        id="«r0»"
        value="This is a text editor"
      />
    </div>
  </div>
</div>
`;
